docs/proposals/scc-check.md

```diff
@@ -0,0 +1,89 @@
+## Problem
+It is impossible for a user to know whether the `PodSpec` he's describing will actually be allowed by the current SCC rules.
+This becomes even more problematic when trying to reason about whether a `DeploymentConfig` or `ReplicationController` or other resource containing a `PodSpec` 
+that a system component will eventually create will work.
+
+## Types
+These types are namespace scoped and only return information about the namespace they're posted to.
+
+```go
+// PodSpecSubjectReview checks whether a particular user/SA tuple can create the PodSpec
+type PodSpecSubjectReview struct {
+	unversioned.TypeMeta `json:",inline"`
+	kapi.ObjectMeta
+
+	Spec PodSpecSubjectReviewSpec `json:"spec"`
+
+	Status PodSpecReviewStatus `json:"status,omitempty"`
+}
+
+type PodSpecSubjectReviewSpec struct {
+	// Spec is the PodSpec to check.
+	PodSpec kapi.PodSpec
+
+	// User is the user you're testing for.
+	// If you specify "User" but not "Group", then is it interpreted as "What if User were not a member of any groups
+	// If User and Groups are empty, then the check is performed using *only* the ServiceAccountName in the PodSpec
+	User string
+	// Groups is the groups you're testing for.
+	Groups []string
+}
+
+type PodSpecReviewStatus struct {
+	// AllowedBy is a reference to the rule that allows the PodSpec.  A `nil`, indicates that it was denied
+	AllowedBy *kapi.ObjectReference `json:"allowedSubjects"`
+
+	// Reason is optional.  It indicates why a request was allowed or denied.
+	Reason string
+
+	// PodSpec is the PodSpec after the defaulting is applied
+	PodSpec kapi.PodSpec
+}
+
+// PodSpecSelfSubjectReview checks whether this user/SA tuple can create the PodSpec
+type PodSpecSelfSubjectReview struct {
+	unversioned.TypeMeta `json:",inline"`
+	kapi.ObjectMeta
+
+	Spec PodSpecSelfSubjectReviewSpec `json:"spec"`
+
+	Status PodSpecReviewStatus `json:"status,omitempty"`
+}
+
+type PodSpecSubjectReviewSpec struct {
+	// Spec is the PodSpec to check.
+	PodSpec kapi.PodSpec
+
+	// ExcludeSelf indicates that the check is performed using *only* the ServiceAccountName in the PodSpec
+	ExcludeSelf bool
```

Comment:
So many words, so little time.

On Wed, Mar 23, 2016 at 1:02 PM, David Eads <notifications@github.com>
wrote:

> In docs/proposals/scc-check.md
> <https://github.com/openshift/origin/pull/8199#discussion_r57196817>:
>
> > +// PodSpecSelfSubjectReview checks whether this user/SA tuple can create the PodSpec
> > +type PodSpecSelfSubjectReview struct {
> > +	unversioned.TypeMeta `json:",inline"`
> > +	kapi.ObjectMeta
> > +
> > +	Spec PodSpecSelfSubjectReviewSpec `json:"spec"`
> > +
> > +	Status PodSpecReviewStatus `json:"status,omitempty"`
> > +}
> > +
> > +type PodSpecSubjectReviewSpec struct {
> > +	// Spec is the PodSpec to check.
> > +	PodSpec kapi.PodSpec
> > +
> > +	// ExcludeSelf indicates that the check is performed using *only* the ServiceAccountName in the PodSpec
> > +	ExcludeSelf bool
>
> Well, can I check for someone else explicitly in one of these?
>
>
> https://github.com/openshift/origin/pull/8199/files/8d2df36a02f307d874ef7da78e123d6508f38edd#diff-3d9f8488bea23bedcf2b8b82b0a52603R20
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8199/files/8d2df36a02f307d874ef7da78e123d6508f38edd#r57196817>
>